x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java

```diff
@@ -123,44 +131,209 @@ public int count() {
             public int get(int i) {
                 return docs.getInt(i);
             }
-        }));
+        };
+        StoredFieldsSpec storedFieldsSpec = StoredFieldsSpec.NO_REQUIREMENTS;
+        List<RowStrideReaderWork> rowStrideReaders = new ArrayList<>(fields.size());
+        try {
+            for (int b = 0; b < fields.size(); b++) {
+                FieldWork field = fields.get(b);
+                BlockLoader.ColumnAtATimeReader columnAtATime = field.columnAtATime.reader(shard, segment, firstDoc);
+                if (columnAtATime != null) {
+                    blocks[b] = (Block) columnAtATime.read(blockFactory, loaderDocs);
+                } else {
+                    BlockLoader.RowStrideReader rowStride = field.rowStride.reader(shard, segment, firstDoc);
+                    rowStrideReaders.add(
+                        new RowStrideReaderWork(
+                            rowStride,
+                            (Block.Builder) field.info.blockLoaders.get(shard).builder(blockFactory, docs.getPositionCount()),
+                            b
+                        )
+                    );
+                    storedFieldsSpec = storedFieldsSpec.merge(field.info.blockLoaders.get(shard).rowStrideStoredFieldSpec());
+                }
+            }
+
+            if (rowStrideReaders.isEmpty()) {
+                return;
+            }
+            BlockLoaderStoredFieldsFromLeafLoader storedFields = null;
+            if (false == storedFieldsSpec.equals(StoredFieldsSpec.NO_REQUIREMENTS)) {
+                storedFields = new BlockLoaderStoredFieldsFromLeafLoader(
+                    // TODO enable the optimization by passing non-null to docs if correct
+                    StoredFieldLoader.fromSpec(storedFieldsSpec).getLoader(ctx(shard, segment), null),
+                    storedFieldsSpec.requiresSource()
+                );
+                trackStoredFields(storedFieldsSpec); // TODO when optimization is enabled add it to tracking
+            }
+            for (int p = 0; p < docs.getPositionCount(); p++) {
+                int doc = docs.getInt(p);
+                if (storedFields != null) {
+                    storedFields.advanceTo(doc);
+                }
+                for (int r = 0; r < rowStrideReaders.size(); r++) {
+                    RowStrideReaderWork work = rowStrideReaders.get(r);
+                    work.reader.read(doc, storedFields, work.builder);
+                }
+            }
+            for (int r = 0; r < rowStrideReaders.size(); r++) {
+                RowStrideReaderWork work = rowStrideReaders.get(r);
+                blocks[work.offset] = work.builder.build();
+            }
+        } finally {
+            Releasables.close(rowStrideReaders);
+        }
     }
 
-    private Block loadFromManyLeaves(DocVector docVector) throws IOException {
+    private void loadFromManyLeaves(Block[] blocks, DocVector docVector) throws IOException {
```

Comment:
Did you intend to always load from stored fields/_source when loading from multiple segments? I think this can be problematic for indices with _source disabled. I also think this may explain the spike in the microbenchmark results for the 'shuffled' section.

Reply:
It loads row-wise, but not using stored fields. The doc values implementations implement both the block-at-a-time interface and the row-stride interface. We just use the row-stride interface there.

I could try to detect when the readers all implement the block-at-a-time interface but that'd need two passes over the shard and segment vectors.